src/action.ts

```diff
@@ -0,0 +1,122 @@
+import { setOutput as setActionOutput } from '@actions/core';
+import semverIncrement from 'semver/functions/inc';
+import semverDiff from 'semver/functions/diff';
+import type { ReleaseType as SemverReleaseType } from 'semver';
+
+import { getTags } from './git-operations';
+import {
+  FieldNames,
+  getMetadataForAllPackages,
+  getPackagesToUpdate,
+  getPackageManifest,
+  PackageManifest,
+  updatePackage,
+  updatePackages,
+} from './package-operations';
+import { getActionInputs, isMajorSemverDiff, WORKSPACE_ROOT } from './utils';
+
+/**
+ * Action entry function. Gets git tags, reads the work space root package.json,
+ * and updates the package(s) of the repository per the Action inputs.
+ *
+ * @see updateMonorepo - For details on monorepo workflow.
+ * @see updatePolyrepo - For details on polyrepo (i.e. single-package
+ * repository) workflow.
+ */
+export async function main(): Promise<void> {
+  const actionInputs = getActionInputs();
+
+  // Get all git tags. An error is thrown if "git tag" returns no tags and the
+  // local git history is incomplete.
+  const [tags] = await getTags();
+
+  const rootManifest = await getPackageManifest(WORKSPACE_ROOT, [
+    FieldNames.Version,
+    FieldNames.Private,
+    FieldNames.Workspaces,
+  ]);
+
+  const { version: currentVersion } = rootManifest;
+  const isMonorepo = rootManifest.private === true && rootManifest.workspaces;
+
+  // Compute the new version and version diff from the inputs and root manifest
+  let newVersion: string, versionDiff: SemverReleaseType;
+  if (actionInputs.ReleaseType) {
+    newVersion = semverIncrement(
+      currentVersion,
+      actionInputs.ReleaseType,
+    ) as string;
+    versionDiff = actionInputs.ReleaseType;
+  } else {
+    newVersion = actionInputs.ReleaseVersion as string;
+    versionDiff = semverDiff(currentVersion, newVersion) as SemverReleaseType;
+  }
+
+  if (isMonorepo) {
+    await updateMonorepo(newVersion, versionDiff, rootManifest, tags);
+  } else {
+    await updatePolyrepo(newVersion, rootManifest);
+  }
+  setActionOutput('NEW_VERSION', newVersion);
+}
+
+/**
+ *
+ * Given that the package is a polyrepo (i.e., a "normal", single-package repo),
+ * updates the package.
+ *
+ * @param newVersion - The package's new version.
+ * @param manifest - The package's parsed package.json file.
+ */
+async function updatePolyrepo(
+  newVersion: string,
+  manifest: Partial<PackageManifest>,
```

Comment:
The package has been validated by this point - I don't think we need to use `Partial` here.
```suggestion
  manifest: PackageManifest,
```

Reply:
I make some changes regarding this in #23. The difficulty lies in `updatePolyrepo` requiring that the root manifest's `name` field exists and is non-empty, while `updateMonorepo` does not. I'd prefer if we deal with it in #23 or a follow-up.